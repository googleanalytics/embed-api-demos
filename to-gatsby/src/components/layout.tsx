// Copyright 2020 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from "react"
import { makeStyles, Theme } from "@material-ui/core/styles"
import Typography from "@material-ui/core/Typography"
import { Link } from "gatsby"
import { Home } from "@material-ui/icons"
import classnames from "classnames"
import Logo from "-!svg-react-loader!../images/ga-developer-logo.svg"
import { useLocation } from "@reach/router"
import AppBar from "@material-ui/core/AppBar"
import IconButton from "@material-ui/core/IconButton"
import Drawer from "@material-ui/core/Drawer"
import List from "@material-ui/core/List"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"
import { navigate } from "@reach/router"
import MenuIcon from "@material-ui/icons/Menu"
import { useSelector } from "react-redux"

import Login from "./Login"

const mobile = (theme: Theme) => theme.breakpoints.between(0, "sm")
const notMobile = (theme: Theme) => theme.breakpoints.up("md")

const useStyles = makeStyles(theme => ({
  root: {
    display: "flex",
    minHeight: "100%",
    [notMobile(theme)]: {
      flexDirection: "row",
    },
    [mobile(theme)]: {
      flexDirection: "column",
    },
  },
  main: {
    display: "flex",
    flexDirection: "column",
    flexGrow: 1,
    minHeight: "100%",
  },
  contentWrapper: {
    flexGrow: 1,
    color: theme.palette.getContrastText(theme.palette.grey[200]),
    backgroundColor: theme.palette.grey[200],
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(4),
    maxWidth: theme.breakpoints.width("md"),
    [mobile(theme)]: {
      maxWidth: "unset",
      width: "100%",
      padding: theme.spacing(2),
    },
  },
  header: {
    padding: theme.spacing(4),
    position: "relative",
    maxWidth: theme.breakpoints.width("md"),
    [mobile(theme)]: {
      maxWidth: "unset",
      width: "100%",
      padding: theme.spacing(2),
    },
  },
  logoRow: {
    display: "flex",
    flexWrap: "wrap",
    alignItems: "center",
    [mobile(theme)]: {
      display: "none",
    },
  },
  logo: {
    flexGrow: 1,
    height: "50px",
  },
  appBarNav: {
    [notMobile(theme)]: {
      display: "none",
    },
    flexDirection: "row",
    alignItems: "center",
    paddingLeft: theme.spacing(1),
  },
  mobileNav: {
    color: theme.palette.getContrastText(theme.palette.grey[800]),
    backgroundColor: theme.palette.grey[800],
    minHeight: "100%",
  },
  nav: {
    [mobile(theme)]: {
      display: "none",
    },
    minWidth: "260px",
    borderRight: `1px solid ${theme.palette.grey[200]}`,
    color: theme.palette.getContrastText(theme.palette.grey[800]),
    backgroundColor: theme.palette.grey[800],
    "& ol": {
      margin: 0,
      padding: 0,
      paddingTop: theme.spacing(1),
      listStyle: "none",
      width: "100%",
      "& li": {
        width: "100%",
        display: "flex",
        "& a": {
          color: "unset",
          width: "100%",
          textDecoration: "none",
        },
      },
    },
  },
  noColor: {
    color: "unset",
  },
  innerNav: {
    padding: theme.spacing(1),
    paddingLeft: theme.spacing(4),
    [mobile(theme)]: {
      paddingLeft: theme.spacing(2),
      paddingRight: theme.spacing(2),
    },
  },
  navLinkBackgroundHover: {
    "&:hover": {
      backgroundColor: theme.palette.grey[100],
      color: theme.palette.getContrastText(theme.palette.grey[100]),
    },
  },
  home: {
    margin: "unset",
    color: "unset",
    marginTop: theme.spacing(1),
    display: "flex",
    alignItems: "center",
    [notMobile(theme)]: {
      "&:hover": {
        color: theme.palette.primary.main,
      },
    },
  },
  subHeading: {
    width: "100%",
    borderTop: `1px solid ${theme.palette.grey[600]}`,
    marginTop: theme.spacing(2),
    paddingTop: theme.spacing(2),
    marginBottom: theme.spacing(1),
    paddingLeft: theme.spacing(4),
    [mobile(theme)]: {
      paddingLeft: theme.spacing(2),
    },
  },
  homeIcon: {
    marginLeft: theme.spacing(-0.5),
    paddingRight: theme.spacing(1),
    fontSize: "1.5em",
  },
  mobileHeading: {
    color: theme.palette.getContrastText(theme.palette.primary.main),
    marginLeft: theme.spacing(1),
  },
  mobileMenu: {
    display: "flex",
    justifyContent: "flex-start",
    flexGrow: 1,
  },
}))

export const usePageView = (
  measurementId: string | undefined = process.env.GATSBY_GA_MEASUREMENT_ID
) => {
  const location = useLocation()
  const gtag = useSelector((a: AppState) => a.gtag)
  React.useEffect(() => {
    if (measurementId === undefined) {
      throw new Error("No measurementId is set.")
    }
    if (gtag === undefined) {
      return
    }
    gtag("config", measurementId, {
      page_path: location.pathname,
    })
  }, [location.pathname, measurementId, gtag])
}

interface LayoutProps {
  title: string
}

interface Heading {
  type: "heading"
  text: string
}

interface LinkT {
  type: "link"
  text: string
  href: string
}

type LinkData = LinkT | Heading

const linkData: LinkData[] = [
  { text: "Demos & Tools", type: "heading" },
  { text: "Autotrack", href: "/autotrack/", type: "link" },
  { text: "Account Explorer", href: "/account-explorer/", type: "link" },
  { text: "Campaign URL Builder", href: "/campaign-url-builder", type: "link" },
  {
    text: "Dimensions & Metrics Explorer",
    href: "/dimensions-metrics-explorer",
    type: "link",
  },
  { text: "Embed API", href: "/embed-api", type: "link" },
  { text: "Enhanced Ecommerce", href: "/enhanced-ecommerce", type: "link" },
  { text: "Hit Builder", href: "/hit-builder", type: "link" },
  { text: "Query Explorer", href: "/query-explorer", type: "link" },
  { text: "Request Composer", href: "/request-composer", type: "link" },
  { text: "Spreadsheet Add-on", href: "/spreadsheet-add-on", type: "link" },
  { text: "Tag Assistant", href: "/tag-assistant", type: "link" },
  { text: "Usage Trends", href: "/usage-trends", type: "link" },
  { text: "Resources", type: "heading" },
  { text: "About this Site", href: "/#about", type: "link" },
  { text: "Help & feedback", href: "/#help", type: "link" },
]

const Layout: React.FC<LayoutProps> = ({ children, title }) => {
  usePageView()
  const classes = useStyles()
  const [open, setOpen] = React.useState(false)

  return (
    <div className={classes.root}>
      <AppBar position="static" className={classes.appBarNav}>
        <IconButton
          edge="start"
          onClick={() => setOpen(true)}
          className={classes.mobileMenu}
        >
          <MenuIcon />
          <Typography variant="h6" className={classes.mobileHeading}>
            GA Demos & Tools
          </Typography>
        </IconButton>
        <Login />
        <Drawer open={open} onClose={() => setOpen(false)}>
          <List className={classes.mobileNav}>
            <Link to="/" className={classes.noColor}>
              <Typography
                className={classnames(classes.innerNav, classes.home)}
                variant="h2"
              >
                <Home className={classes.homeIcon} /> Home
              </Typography>
            </Link>
            {linkData.map(linkData => {
              if (linkData.type === "heading") {
                return (
                  <Typography
                    key={linkData.text}
                    className={classnames(classes.subHeading, classes.innerNav)}
                    variant="h6"
                  >
                    {linkData.text}
                  </Typography>
                )
              }
              return (
                <ListItem
                  button
                  key={linkData.text}
                  className={classes.innerNav}
                  onClick={() => {
                    setOpen(false)
                    navigate(linkData.href)
                  }}
                >
                  {linkData.text}
                </ListItem>
              )
            })}
          </List>
        </Drawer>
      </AppBar>
      <nav className={classes.nav}>
        <ol>
          <li>
            <Link to="/">
              <Typography
                className={classnames(classes.innerNav, classes.home)}
                variant="h2"
              >
                <Home className={classes.homeIcon} /> Home
              </Typography>
            </Link>
          </li>
          {linkData.map(linkData => {
            if (linkData.type === "heading") {
              return (
                <li key={linkData.text}>
                  <Typography
                    className={classnames(classes.subHeading)}
                    variant="h6"
                  >
                    {linkData.text}
                  </Typography>
                </li>
              )
            }
            return (
              <li key={linkData.text}>
                <Link
                  className={classnames(
                    classes.innerNav,
                    classes.navLinkBackgroundHover
                  )}
                  to={linkData.href}
                >
                  {linkData.text}
                </Link>
              </li>
            )
          })}
        </ol>
      </nav>
      <main className={classes.main}>
        <header className={classes.header}>
          {/* TODO - Figure out how to size the logo correctly. I probably want to use media queries with useStyles() */}
          <div className={classes.logoRow}>
            <Logo className={classes.logo} />
            <Login />
          </div>
          <Typography variant="h1">{title}</Typography>
        </header>
        <div className={classes.contentWrapper}>
          <section className={classes.content}>{children}</section>
        </div>
      </main>
    </div>
  )
}

export default Layout
