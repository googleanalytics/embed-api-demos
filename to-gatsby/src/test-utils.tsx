// Copyright 2020 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from "react"
import { Provider } from "react-redux"
import { store } from "../gatsby/wrapRootElement"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
  History,
} from "@reach/router"

export const withProviders = (
  component: JSX.Element,
  path: string = "/"
): {
  wrapped: JSX.Element
  history: History
} => {
  const history = createHistory(createMemorySource(path))
  const wrapped = (
    <Provider store={store}>
      <LocationProvider history={history}>{component}</LocationProvider>
    </Provider>
  )
  return { wrapped, history }
}
