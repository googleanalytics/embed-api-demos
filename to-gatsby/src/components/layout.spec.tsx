// Copyright 2020 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from "react"
import * as renderer from "react-test-renderer"
import { withProviders } from "../test-utils"

import Layout from "./layout"
import { usePageView } from "./layout"

const TestMeasurementId = "abc"

beforeEach(() => {
  process.env.GATSBY_GA_MEASUREMENT_ID = TestMeasurementId
  window.gtag = jest.fn()
})

describe("Layout", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        withProviders(<Layout title="Page Title">Content</Layout>).wrapped
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe("usePageView hook", () => {
  describe("when gtag is undefined", () => {
    test("and stays that way doesn't rendered forever", () => {
      jest.useFakeTimers()
      const hookValues: ReturnType<typeof usePageView>[] = []
      ;(window.gtag as any) = undefined
      const TestComponent: React.FC = () => {
        const runData = usePageView()
        hookValues.push(runData)
        return null
      }
      const component = withProviders(<TestComponent />).wrapped
      renderer.create(component)
      // advance the timer more times than necessary so we can make sure this runs
      // enough times.
      for (let i = 0; i < 30; i++) {
        renderer.act(() => {
          jest.advanceTimersByTime(40)
        })
      }
      // Make sure that it was never able to measure since gtag was undefined
      // the entire time.
      expect(hookValues.every(a => a.couldMeasure === false))
      // Make sure that the hook tried to run at least a few times before giving up.
      expect(hookValues.length).toBeGreaterThan(1)
    })
    test("but is defined later", () => {
      const gtagMock = jest.fn()
      const pathOverride = "/hello"
      jest.useFakeTimers()
      const capturedValues: any[] = []
      ;(window.gtag as any) = undefined
      const TestComponent: React.FC = () => {
        const runData = usePageView()
        capturedValues.push(runData)
        return null
      }
      const component = withProviders(<TestComponent />, pathOverride).wrapped

      renderer.create(component)
      renderer.act(() => {
        jest.advanceTimersByTime(100)
      })
      window.gtag = gtagMock
      renderer.act(() => {
        jest.advanceTimersByTime(100)
      })

      // Make sure that couldMeasure was eventually true.
      expect(capturedValues.every(a => a.couldMeasure === false)).toBeFalsy()
      const gtagCalls = gtagMock.mock.calls
      expect(gtagCalls).toHaveLength(1)
      expect(gtagCalls[0]).toEqual([
        "config",
        TestMeasurementId,
        { page_path: pathOverride },
      ])
    })
  })
  describe("when gtag is defined", () => {
    test("and user navigates to page", async () => {
      const gtagMock = jest.fn()
      window.gtag = gtagMock
      const path1 = "/hello"
      const path2 = "/second"
      const TestComponent: React.FC = () => {
        usePageView()
        return null
      }
      const { wrapped: component, history } = withProviders(
        <TestComponent />,
        path1
      )

      renderer.create(component)
      await renderer.act(async () => {
        // Navigate to another page
        await history.navigate(path2)
      })
      const gtagCalls = gtagMock.mock.calls
      expect(gtagCalls).toHaveLength(2)
      expect(gtagCalls[0]).toEqual([
        "config",
        TestMeasurementId,
        { page_path: path1 },
      ])
      expect(gtagCalls[1]).toEqual([
        "config",
        TestMeasurementId,
        { page_path: path2 },
      ])
    })
  })
})
