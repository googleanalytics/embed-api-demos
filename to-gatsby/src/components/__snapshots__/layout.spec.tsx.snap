// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-appBarNav-8 MuiPaper-elevation4"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root makeStyles-mobileMenu-18 MuiIconButton-edgeStart"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <h6
          className="MuiTypography-root makeStyles-mobileHeading-17 MuiTypography-h6"
        >
          GA Demos & Tools
        </h6>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-describedby={null}
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      title="Login"
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5.9c1.16 0 2.1.94 2.1 2.1s-.94 2.1-2.1 2.1S9.9 9.16 9.9 8s.94-2.1 2.1-2.1m0 9c2.97 0 6.1 1.46 6.1 2.1v1.1H5.9V17c0-.64 3.13-2.1 6.1-2.1M12 4C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 9c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </header>
  <nav
    className="makeStyles-nav-10"
  >
    <ol>
      <li>
        <a
          href="/"
        >
          <h2
            className="MuiTypography-root makeStyles-innerNav-12 makeStyles-home-14 MuiTypography-h2"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root makeStyles-homeIcon-16"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
             Home
          </h2>
        </a>
      </li>
      <li>
        <h6
          className="MuiTypography-root makeStyles-subHeading-15 MuiTypography-h6"
        >
          Demos & Tools
        </h6>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/autotrack/"
        >
          Autotrack
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/account-explorer/"
        >
          Account Explorer
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/campaign-url-builder"
        >
          Campaign URL Builder
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/dimensions-metrics-explorer"
        >
          Dimensions & Metrics Explorer
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/embed-api"
        >
          Embed API
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/enhanced-ecommerce"
        >
          Enhanced Ecommerce
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/hit-builder"
        >
          Hit Builder
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/query-explorer"
        >
          Query Explorer
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/request-composer"
        >
          Request Composer
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/spreadsheet-add-on"
        >
          Spreadsheet Add-on
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/tag-assistant"
        >
          Tag Assistant
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/usage-trends"
        >
          Usage Trends
        </a>
      </li>
      <li>
        <h6
          className="MuiTypography-root makeStyles-subHeading-15 MuiTypography-h6"
        >
          Resources
        </h6>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/#about"
        >
          About this Site
        </a>
      </li>
      <li>
        <a
          className="makeStyles-innerNav-12 makeStyles-navLinkBackgroundHover-13"
          href="/#help"
        >
          Help & feedback
        </a>
      </li>
    </ol>
  </nav>
  <main
    className="makeStyles-main-2"
  >
    <header
      className="makeStyles-header-5"
    >
      <div
        className="makeStyles-logoRow-6"
      >
        <test-file-stub
          className="makeStyles-logo-7"
        />
        <button
          aria-describedby={null}
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Login"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5.9c1.16 0 2.1.94 2.1 2.1s-.94 2.1-2.1 2.1S9.9 9.16 9.9 8s.94-2.1 2.1-2.1m0 9c2.97 0 6.1 1.46 6.1 2.1v1.1H5.9V17c0-.64 3.13-2.1 6.1-2.1M12 4C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 9c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
      <h1
        className="MuiTypography-root MuiTypography-h1"
      >
        Page Title
      </h1>
    </header>
    <div
      className="makeStyles-contentWrapper-3"
    >
      <section
        className="makeStyles-content-4"
      >
        Content
      </section>
    </div>
  </main>
</div>
`;
